bin.src/generateCommands.py

```diff
@@ -0,0 +1,581 @@
+        if not args.force:
+            sys.exit(1)
+
+    unrecognisedBlocks = set(args.blocks) - set(possibleBlocks)
+    if unrecognisedBlocks:
+        print(f"Unrecognised blocks: '{list(unrecognisedBlocks)}'%s" % (" ignoring" if args.force else ""),
+              file=sys.stderr)
+        if args.verbose:
+            print(f"Possible blocks are {list(possibleBlocks)}", file=sys.stderr)
+        if not args.force:
+            sys.exit(1)
+
+    #
+    # Bootstrap the calibrations
+    #
+    if args.bootstrap:
+        bootstrap.execute(
+            args.dataDir, args.calib, args.copyMode,
+            verbose=args.verbose
+            )
+
+    #
+    # Process the main blocks
+    #
+
+    for blockName in args.blocks:
+        if blockName in calibBlocks:
+            calibBlocks[blockName].execute(
+                args.dataDir, args.calib, args.rerun, args.copyMode,
+                calibTypes=args.calibTypes, processes=args.processes, verbose=args.verbose
+                )
+
+    for blockName in args.blocks:
+        if blockName in scienceBlocks:
+            scienceBlocks[blockName].execute(
+                args.dataDir, args.calib, args.rerun,
+                steps=args.scienceSteps, processes=args.processes, verbose=args.verbose
+                )
+
+
+class BootstrapSource:
+    __slots__ = ["dirName", "detectorMapFmt", "arms"]
+
+    def __init__(self, dirName:str, detectorMapFmt:str, arms:Iterator[str]):
+        self.dirName        = str(dirName)
+        self.detectorMapFmt = str(detectorMapFmt)
+        self.arms           = list(arms)
+
+    @classmethod
+    def fromYaml(cls, yamlBlock:Mapping[str,Any]) -> "BootstrapSource":
+        r"""
+        \param yamlBlock: {
+            "dirName": "$DRP_STELLA_DATA_DIR/raw",
+            "detectorMapFmt": "detectorMap-sim-%s.fits",
+            "arms": ["r1", "b1"],
+            }
+        """
+        return cls(yamlBlock["dirName"], yamlBlock["detectorMapFmt"], yamlBlock["arms"])
+
+    def execute(self, dataDir, calib, copyMode, verbose=False):
+        bootstrapDir = os.path.join(dataDir, os.path.expandvars(self.dirName))
+        if verbose:
+            print("Reading bootstrap files from %s" % (bootstrapDir), file=sys.stderr)
+
+        detectorMaps = []
+        for arm in self.arms:
+            detectorMaps.append(os.path.join(bootstrapDir, self.detectorMapFmt % arm))
+
+        command = [
+            "ingestCalibs.py",
+            dataDir,
+            f"--output={os.path.join(dataDir, calib)}",
+            "--validity=1800",
+            "--create",
+            f"--mode={copyMode}",
+            "--",
+            ]
+
+        command += detectorMaps
+
+        print(f"{shellCommand(command)} || return 1")
+
+
+class CalibSource:
+    __slots__ = ["configs", "visits"]
+    __subclasses:Dict[str,type] = {}
+
+    def __init__(self, configs:Iterator[str], visits:Iterator[int]):
+        self.configs = list(configs)
+        self.visits  = list(visits)
```

Comment:
Why does it have to take a list of visits? We can be so much more expressive about how to select visits than just giving a number, e.g., `--id field=BIAS dateObs=2020-05-13`. I would hate to lose this capability!

Reply:
Made `field` available. It is easy to add other keys like `dateObs`, but I have not done it yet.